README.txt
= Reststop

<b>Reststop makes it easy to write RESTful[http://en.wikipedia.org/wiki/Representational_State_Transfer] 
applications in Camping[http://camping.rubyforge.org/files/README.html].</b>

For info and downloads please see http://rubyforge.org/projects/reststop


*Author*::    Matt Zukowski (matt at roughest dot net)
*Copyright*:: Copyright (c) 2007 Urbacon Ltd.
*License*::   GNU Lesser General Public License Version 3


For an example of a complete Reststop-based Camping app, have a look at 
http://reststop.rubyforge.org/svn/trunk/examples/blog.rb

Reststop essentially gives you three things:

<b>1. Camping controllers that respond to the standard REST verbs:</b>

* create (POST)
* read (GET)
* update (PUT)
* destroy (DELETE)
* list (GET)

Custom actions are also possible. See the Camping::Controllers#REST method documentation for usage info.

<b>2. Camping views grouped by output format:</b>

Your views module:

  module Foobar::Views
    module HTML
      def foo
        html do
          p "Hello World"
        end
      end
    end
    module XML
      def foo
        tag!('foo')
          "Hello World"
        end
      end
    end
  end

Your render call:

  render(:foo, :XML)
  
See the Camping#render method documentation for usage info.

<b>3. Nice URLs to bring it all together:</b>

For example a list of kittens in the default format (HTML) is available at:

  /kittens

The list, in RSS format:

  /kittens.rss
  
Kitten with id 1, in XML format:
  
  /kittens/1.xml
  
Using custom action 'meow' on kitten with id 1:

  /kittens/1/meow

In other words, say you have a "kittens" resource; you can make a GET
request to http://yourapp.com/kittens.xml and get a list of kittens
through your Kittens controller's <tt>list</tt>, formatted using your
<tt>XML</tt> view module. 


<b>Looking for Restr, the very simple REST client?</b>

Due to its popularity, <b>Restr</b>, the "very simple rest client",
has been moved out to its own gem. Restr now lives under
http://rubyforge.org/projects/restr/ and can be installed using:

  gem install restr
 
Restr is basically a wrapper around Ruby's Net::HTTP, offering
a more RESTfully meaningful interface.

See the "Restr documentation"[http://restr.rubyforge.org/] for more info, 
but here's a simple example of RESTful interaction with Restr:

  require 'restr'
  kitten = Restr.get('http://example.com/kittens/1.xml')
  puts kitten['name']
  puts kitten['colour']
  
  kitten['colour'] = 'black'
  kitten = Restr.put('http://example.com/kittens/1.xml', kitten)


== License

Reststop is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published 
by the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

Reststop is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

History.txt
=== 0.5.3 :: 2010-07-07 
  * Fixed issue in the new basic_render method where Tilt templates were not supported (on either actions or layout)

=== 0.5.2 :: 2010-06-21 
  * Made the content-type in reststop_render to be text/html only if the format is :HTML, and application/#{format} otherwise
  * Fixed issue in the new basic_render method where the layout was not being applied.

=== 0.5.1 :: 2010-06-09 
  * Update to align with the post 2.0 enhancements in Camping related to Tilt support.
	Since Tilt support is now checking for the existing of view methods and Tilt views
	RESTstop needs to ensure that view methods can be looked up in Views:HTML as opposed to Views.

=== 0.5.0 :: 2010-04-07

  * Major refactoring for compatibility with Camping 2.0. Thanks to Philippe Monnet
    for completing the work.

=== 0.4.1 :: 2009-05-08

  * Custom actions should now work again (needed to make some more changes for
    Camping 2.0 compatibility).

=== 0.4.0 :: 2009-05-01

  * Now compatible with Camping 2.0, but no longer works with Camping 1.x :(

=== 0.3.0 :: 2008-06-19

  * Restr has been moved out to its own gem, and should now work properly with 
    Rails 2.x (fixes bug #17421).
  * `R` helper method should now work properly when a Restful controlelr class 
    is fed as input along with a custom action. The final argument can be a Hash 
    and will be translated into URL parameters. For example: 
      R(Kittens, 'list', 'colour' => 'black') 

=== 0.2.0 :: 2007-12-10

  * It is now possible to specify the format for rendering a page by appending
    it as a filename extension to the URL. For example: '/items.xml' to use the XML
    view module, or '/items/5.rss' to use the RSS module.
  * Custom actions are now possible. Custom methods defined in your controller will 
    respond to the standard custom-action URLs. For example, if you have a method
    'foo' in your 'kittens' controller, you can get to it by requesting the URL
    '/kittens/foo' or '/kittens/1/foo'. Note that the action will respond to any
    HTTP method (post, get, put, delete). It is up to you to handle different
    request methods inside your action's definition.
  * XML input (from an ActiveResource client, for example) is now parsed into a
    nested Hash, and is made available as the standard @input Camping variable.
  * HTML forms created using Markaby (i.e. by just using `form`) that
    have a :method parameter now automatically insert a hidden '_method' input to
    facilitate 'put' and 'delete' HTTP methods for HTML browsers.
  * Pages now correctly return HTTP status code 501 when an unimplemented controller 
    method is called. ActiveRecord::NotFound exceptions are caught and return 404 
    status.
  * R helper now works for routing RESTful controllers.

=== 0.1.0 :: 2007-07-30

  * First public release.


test/test.rb
#	Prerequisites:
#		1) 	gem install xml-simple
#		2)		gem install restr

gem 'xml-simple'
gem 'restr'
require 'restr'

logger = Logger.new('restr.log')
logger.level = Logger::DEBUG
Restr.logger = logger

u0 =  "http://localhost:3301/sessions.xml"
o = { :username=>'admin', :password=>'camping'}
p0=Restr.post(u0,o)

u1 = "http://localhost:3301/posts/1.xml"
p = Restr.get(u1,o)

# Modify the title
p['title']='HOT off the presses: ' + p['title']

# Update the resource
p2=Restr.put(u1,p,o)

u3="http://localhost:3301/posts.xml"
p3={ :title=>'Brand new REST-issued post', :body=>'RESTstop makes it happen!!!'} 
p4=Restr.post(u2,p3)

u3="http://localhost:3301/posts/4.xml"
p5=Restr.delete(u3)

test/test_helper.rb
require 'test/unit'
require File.dirname(__FILE__) + '/../lib/reststop'


test/reststop_test.rb
require File.dirname(__FILE__) + '/test_helper.rb'

class ReststopTest < Test::Unit::TestCase

  def setup
  end
  
  def test_truth
    assert true
  end
end


lib/restr.rb
require 'net/http'
require 'net/https'
require 'uri'
require 'cgi'

begin
  require 'xml_simple'
rescue LoadError
  require 'rubygems'
  require 'active_support'
  begin
  require 'xml_simple'
  rescue LoadError
    require 'xmlsimple'
  end
end


# A very simple REST client, best explained by example:
#
#   # Retrieve a Kitten and print its name and colour
#   kitten = Restr.get('http://example.com/kittens/1.xml')
#   puts kitten['name']
#   puts kitten['colour']
#
#   # Create a Kitten
#   kitten = Restr.post('http://example.com/kittens.xml', 
#     :name => 'batman', :colour => 'black')
#
#   # Update a Kitten
#   kitten = Restr.put('http://example.com/kittens/1.xml', 
#     :age => '6 months')
#
#   # Delete a Kitten :(
#   kitten = Restr.delete('http://example.com/kittens/1.xml')
#
#   # Retrieve a list of Kittens
#   kittens = Restr.get('http://example.com/kittens.xml')
#
# When the response to a Restr request has content type 'text/xml', the
# response body will be parsed from XML into a nested Hash (using XmlSimple 
# -- see http://xml-simple.rubyforge.org/). Otherwise the response is  
# returned untouched, as a String.
#
# If the remote REST resource requires authentication (Restr only supports
# HTTP Basic authentication, for now):
#
#   Restr.get('http://example.com/kittens/1.xml, {}, 
#     {:username => 'foo', :password => 'bar'})
#
class Restr
  @@log = nil
  
  def self.logger=(logger)
    @@log = logger.dup
    # ActiveSupport's BufferedLogger doesn't seem to support progname= :(
    @@log.progname = self.name if @@log.respond_to?(:progname)
  end
  
  def self.method_missing(method, *args)
    self.do(method, args[0], args[1] || {}, args[2])
  end
  
  def self.do(method, url, params = {}, auth = nil)
    uri = URI.parse(url)
    params = {} unless params
      
    method_mod = method.to_s.downcase.capitalize
    unless Net::HTTP.const_defined?(method_mod)
      raise InvalidRequestMethod, 
        "Callback method #{method.inspect} is not a valid HTTP request method."
    end
    
    if method_mod == 'Get'
      q = params.collect{|k,v| "#{CGI.escape(k.to_s)}=#{CGI.escape(v.to_s)}"}.join("&")
      if uri.query
        uri.query += "&#{q}"
      else
        uri.query = q
      end
    end
    
    req = Net::HTTP.const_get(method_mod).new(uri.request_uri)
    
    
    if auth
      raise ArgumentError, 
        "The `auth` parameter must be a Hash with a :username and :password value." unless 
        auth.kind_of? Hash
      req.basic_auth auth[:username] || auth['username'], auth[:password] || auth['password']
    end
    
    unless method_mod == 'Get'
      req.set_form_data(params, '&')
    end
    
    @@log.debug("Sending #{method.inspect} request to #{url.inspect} with data #{params.inspect}"+
        (auth ? " with authentication" : "")+".") if @@log
 
    client = Net::HTTP.new(uri.host, uri.port)
    client.use_ssl = (uri.scheme == 'https')
    res = client.start do |http|
      http.request(req)
    end
    
    case res
    when Net::HTTPSuccess
      if res.content_type == 'text/xml'
        @@log.debug("Got XML response.") if @@log
        return XmlSimple.xml_in_string(res.body,
          'forcearray'   => false,
          'keeproot'     => false
        )
      else
        @@log.debug("Got #{res.content_type.inspect} response.") if @@log
        return res.body
      end
    else
      $LAST_ERROR_BODY = res.body # FIXME: this is dumb... need a better way of reporting errors
      @@log.error("Got error response '#{res.message}(#{res.code})': #{$LAST_ERROR_BODY}") if @@log
      res.error!      
    end
  end
  
  class InvalidRequestMethod < Exception
  end
end

examples/blog.rb
#!/usr/bin/env ruby

require 'rubygems'
#require 'ruby-debug'			# @techarch : commented out since only needed for local debugging

require 'markaby'					# @techarch : added explicit require
require 'camping'					# @techarch
require 'camping/session'	# @techarch : added explicit require since session has changed in Camping 2.0

gem 'RedCloth'						# @techarch : added since it is referenced in the Posts model
require 'redcloth'				# @techarch : added

#gem 'camping', '~> 2.0'
gem 'camping' , '>= 2.0'	# @techarch : updated version

#gem 'reststop', '~> 0.3'

=begin										# @techarch : commented out since only needed for local debugging
$: << '../../camping/lib'
$: << '../lib'
require 'camping-unabridged'
require 'camping/ar'
require 'camping/session'
=end

#begin										# @techarch : commented out since only needed for local debugging
  # try to use local copy of library
  #require '../lib/reststop2'
  $: << '../lib/'
  require 'reststop.rb'			# @techarch : adjusted so that it is located in the same current folder
#rescue LoadError
#  # ... otherwise default to rubygem
#  require 'reststop'
#end

Camping.goes :Blog

module Blog
  include Camping::Session
  include Reststop
 
  Controllers.extend Reststop::Controllers
end

module Blog::Base
  alias camping_render render
  alias camping_lookup lookup	# @techarch: required if camping > 2.0
  alias camping_service service
  include Reststop::Base
  alias service reststop_service
  alias render reststop_render
  
	# Overrides the new Tilt-centric lookup method In camping
	# RESTstop needs to have a first try at looking up the view
	# located in the Views::HTML module. 
    def lookup(n)
      T.fetch(n.to_sym) do |k|
        t = Blog::Views::HTML.method_defined?(k) || camping_lookup(n)
      end
    end
end

module Blog::Models
  class Post < Base
    belongs_to :user

    before_save do |record|
       cloth = RedCloth.new(record.body)
       cloth.hard_breaks = false
       record.html_body = cloth.to_html
    end
  end

  class Comment < Base; belongs_to :user; end
  class User < Base; end

  class BasicFields < V 1.1
    def self.up
      create_table :blog_posts, :force => true do |t|
        t.integer :user_id,          :null => false
        t.string  :title,            :limit => 255
        t.text    :body, :html_body
        t.timestamps
      end
      create_table :blog_users, :force => true do |t|
        t.string  :username, :password
      end
      create_table :blog_comments, :force => true do |t|
        t.integer :post_id,          :null => false
        t.string  :username
        t.text    :body, :html_body
        t.timestamps
      end
      User.create :username => 'admin', :password => 'camping'
    end

    def self.down
      drop_table :blog_posts
      drop_table :blog_users
      drop_table :blog_comments
    end
  end
end

module Blog::Controllers
  extend Reststop::Controllers
    class Index
		def get
			redirect '/posts'
		end
	end
	
	class Login < R '/login'		# @techarch : added explicit login controller	
		def get
			render :_login
		end
	end
	
    class Posts < REST 'posts'      
      # POST /posts
      def create
        require_login!
        @post = Post.create :title => (input.post_title || input.title),	# @techarch : allow for REST-client based update 
		  :body => (input.post_body || input.body),								# @techarch : allow for REST-client based update
          :user_id => @state.user_id
        redirect R(@post)	
      end

      # GET /posts/1
      # GET /posts/1.xml
      def read(post_id)
        @post = Post.find(post_id)
        @comments = Models::Comment.find(:all, :conditions => ['post_id = ?', post_id])
        render :view
      end

      # PUT /posts/1
      def update(post_id)
        require_login!
        @post = Post.find(post_id)
        @post.update_attributes :title => (input.post_title || input.title),	# @techarch : allow for REST-client based update 
			:body => (input.post_body || input.body)									# @techarch : allow for REST-client based update 
        redirect R(@post)
      end

      # DELETE /posts/1
      def delete(post_id)
        require_login!
        @post = Post.find post_id

        if @post.destroy
          redirect R(Posts)
        else
          _error("Unable to delete post #{@post.id}", 500)
        end
      end

      # GET /posts
      # GET /posts.xml
      def list
        @posts = Post.all(:order => 'updated_at DESC')
        s=render :index
		s
      end

      # GET /posts/new
      def new
        #@state.user_id = 1		# @techarch : commented out as was probably hard-coded for testing purpose
        require_login!
		@user = User.find @state.user_id	# @techarch : added since we need the user info
        @post = Post.new
        render :add
      end

      # GET /posts/1/edit
      def edit(post_id)
        require_login!
 		@user = User.find @state.user_id	# @techarch : added since we need the user info
        @post = Post.find(post_id)
        render :edit
      end
    end
     
    class Comments < REST 'comments'
      # POST /comments
      def create
        Models::Comment.create(:username => (input.post_username || input.username),	# @techarch : allow for REST-client based		update 
			:body => (input.post_body || input.body),	# @techarch : allow for REST-client based update  
			:post_id => input.post_id)
        redirect R(Posts, input.post_id)
      end
    end

    class Sessions < REST 'sessions'
        # POST /sessions
        def create
          @user = User.find_by_username_and_password(input.username, input.password)

          if @user
            @state.user_id = @user.id
            redirect R(Posts)
          else
            @info = 'Wrong username or password.'
          end
          render :login
        end   

        # DELETE /sessions
        def delete
          @state.user_id = nil
          redirect R(Posts) 		# @techarch : changed redirect from Index (does not exist) to Posts
        end
    end
    
    # You can use old-fashioned Camping controllers too!
    class Style < R '/styles.css'
      def get
        @headers["Content-Type"] = "text/css; charset=utf-8"
        @body = %{
          body {
              font-family: Utopia, Georga, serif;
          }
          h1.header {
              background-color: #fef;
              margin: 0; padding: 10px;
          }
          div.content {
              padding: 10px;
          }
        }
      end
    end
end

module Blog::Helpers
  alias_method :_R, :R
  remove_method :R
  include Reststop::Helpers

  def logged_in?
    !!@state.user_id
  end

  def require_login!
    unless logged_in?
      redirect(R(Blog::Controllers::Login))	# @techarch: add explicit route
      throw :halt
    end
  end
end


module Blog::Views
  extend Reststop::Views
  
  module HTML
    include Blog::Controllers
    include Blog::Views

    def layout
      html do
        head do
          title 'blog'
          link :rel => 'stylesheet', :type => 'text/css',
               :href => self/'/styles.css', :media => 'screen'
        end
        body do
          h1.header { a 'blog', :href => R(Posts) }
          div.content do
            self << yield
          end
        end
      end
    end

    def index
      if @posts.empty?
        p 'No posts found.'
      else
        for post in @posts
          _post(post)
        end
      end
      p { a 'Add', :href => R(Posts, 'new') }
    end

    def login
      p { b @login }
      p { a 'Continue', :href => R(Posts, 'new') }
    end

    def logout
      p "You have been logged out."
      p { a 'Continue', :href => R(Posts) }
    end

    def add
      if @user
        _form(@post, :action => R(Posts))
      else
        _login
      end
    end

    def edit
      if @user
        _form(@post, :action => R(@post), :method => :put) 
      else
        _login
      end
    end

    def view
      _post(@post)

      p "Comment for this post:"
      for c in @comments
        h1 c.username
        p c.body
      end

      form :action => R(Comments), :method => 'post' do
        label 'Name', :for => 'post_username'; br
        input :name => 'post_username', :type => 'text'; br
        label 'Comment', :for => 'post_body'; br
        textarea :name => 'post_body' do; end; br
        input :type => 'hidden', :name => 'post_id', :value => @post.id
        input :type => 'submit'
      end
    end

    # partials
    def _login
      p do
        "(default: admin/camping)"
      end
      form :action => R(Sessions), :method => 'post' do
        label 'Username', :for => 'username'; br
        input :name => 'username', :type => 'text'; br

        label 'Password', :for => 'password'; br
        input :name => 'password', :type => 'text'; br

        input :type => 'submit', :name => 'login', :value => 'Login'
      end
    end

    def _post(post)
      h1 post.title
      p post.body
      p do
        [a("Edit", :href => R(Posts, post.id, 'edit')), a("View", :href => R(Posts, post.id, 'edit'))].join " | "
      end
    end

    def _form(post, opts)
      form(:action => R(Sessions), :method => 'delete') do
      p do
        span "You are logged in as #{@user.username}"
        span " | "
        button(:type => 'submit') {'Logout'}
      end
      end
      form({:method => 'post'}.merge(opts)) do
        label 'Title', :for => 'post_title'; br
        input :name => 'post_title', :type => 'text',
              :value => post.title; br

        label 'Body', :for => 'post_body'; br
        textarea post.body, :name => 'post_body'; br

        input :type => 'hidden', :name => 'post_id', :value => post.id
        input :type => 'submit'
      end
    end
  end
  default_format :HTML

  module XML
    def layout
      yield
    end

    def index
      @posts.to_xml(:root => 'blog')
    end

    def view
      @post.to_xml(:root => 'post')
    end
  end
end
 
def Blog.create
  raise "You must configure the database first in 'config/database.yml'!" unless File.exist?('config/database.yml')
 	dbconfig = YAML.load(File.read('config/database.yml'))								# @techarch
	Camping::Models::Base.establish_connection  dbconfig['development']		# @techarch

	Blog::Models.create_schema :assume => (Blog::Models::Post.table_exists? ? 1.0 : 0.0)
end


lib/reststop.rb
# Right now you'll have to do some weird gymnastics to get this hooked in to a Camping app...
# Something like:
#
#   Camping.goes :Blog
# 
#   module Blog
#     include Camping::Session
#     include Reststop
#  
#     Controllers.extend Reststop::Controllers
#   end
# 
#   module Blog::Base
#     alias camping_render render
#     alias camping_service service
#		alias camping_lookup lookup
#     include Reststop::Base
#     alias service reststop_service
#     alias render reststop_render
#
#	# Overrides the new Tilt-centric lookup method In camping
#	# RESTstop needs to have a first try at looking up the view
#	# located in the Views::HTML module. 
#   def lookup(n)
#      T.fetch(n.to_sym) do |k|
#        t = Blog::Views::HTML.method_defined?(k) || camping_lookup(n)
#      end
#    end

#   end
#   
#   module Blog::Controllers
#     extend Reststop::Controllers
#     ...
#   end
#   
#   module Blog::Helpers
#     alias_method :_R, :R
#     remove_method :R
#     include Reststop::Helpers
#     ...
#   end
#   
#   module Blog::Views
#     extend Reststop::Views
#     ...
#   end
#
# The hope is that this could all get taken care of in a
# `include Reststop` call (via overriding of #extended)
#
# See examples/blog.rb for a working example.
require 'logger'
$LOG = Logger.new(STDOUT)

module Reststop
  module Base
    def reststop_service(*a)
      if @env['REQUEST_METHOD'] == 'POST' && (input['_method'] == 'put' || input['_method'] == 'delete')
        @env['REQUEST_METHOD'] = input._method.upcase
        @method = input._method
      end
  	  @a0=a[0] if !a.empty?
      camping_service(*a)
    end

    # Overrides Camping's render method to add the ability to specify a format
    # module when rendering a view.
    #
    # The format can also be specified in other ways (shown in this order
    # of precedence):
    #
    # 1. By providing a second parameter to render()
    #    (eg: <tt>render(:foo, :HTML)</tt>)
    # 2. By setting the @format variable
    # 3. By providing a 'format' parameter in the request (i.e. input[:format])
    # 4. By adding a file-format extension to the url (e.g. /items.xml or
    #    /items/2.html).
    #
    # For example, you could have:
    #
    #   module Foobar::Views
    #
    #     module HTML
    #       def foo
    #         # ... render some HTML content
    #       end
    #     end
    #
    #     module RSS
    #       def foo
    #         # ... render some RSS content
    #       end
    #     end
    #
    #   end
    #
    # Then in your controller, you would call render() like this:
    #
    #   render(:foo, :HTML) # render the HTML version of foo
    #
    # or
    #
    #   render(:foo, :RSS) # render the RSS version of foo
    #
    # or
    #
    #   @format = :RSS
    #   render(:foo) # render the RSS version of foo
    #
    # or
    #
    #   # url is /foobar/1?format=RSS
    #   render(:foo) # render the RSS version of foo
    #
    # or
    #
    #   # url is /foobar/1.rss
    #   render(:foo) # render the RSS version of foo
    #
    # If no format is specified, render() will behave like it normally does in
    # Camping, by looking for a matching view method directly
    # in the Views module.
    #
    # You can also specify a default format module by calling
    # <tt>default_format</tt> after the format module definition.
    # For example:
    #
    #   module Foobar::Views
    #     module HTML
    #       # ... etc.
    #     end
    #     default_format :HTML
    #   end
    #
    def reststop_render(action, format = nil)
	  format = nil unless format.is_a? Symbol

	  app_name = self.class.name.split("::").first							# @techarch : get the name of the app
	  format ||= @format

      if format.nil?
        begin
          ct = CONTENT_TYPE
        rescue NameError
          ct = 'text/html'
        end

        @headers['Content-Type'] ||= ct
   		basic_render(action) # @techarch
      else
        mab = (app_name + '::Mab').constantize								# @techarch : get the Mab class
    	m = mab.new({}, self)																# @techarch : instantiate Mab
    	mod = (app_name + "::Views::#{format.to_s}").constantize	# @techarch : get the right Views format class

        m.extend mod

        begin
          ct = mod::CONTENT_TYPE
        rescue NameError
          ct = "#{format == :HTML ? 'text' : 'application'}/#{format.to_s.downcase}" #@techarch - used to be text/***
        end
        @headers['Content-Type'] = ct

        s = m.capture{m.send(action)}
        s = m.capture{send(:layout){s}} if /^_/!~action.to_s and m.respond_to?(:layout)	
        s
      end
    end
	
	# Performs a basic camping rendering (without use of a layout)
	# This method was added since the addition of Tilt support in camping 
	# is assuming layout.
	def basic_render(action)
		app_name = self.class.name.split("::").first							# @techarch : get the name of the app
        	mab = (app_name + '::Mab').constantize								# @techarch : get the Mab class
   		m = mab.new({}, self)														# @techarch : instantiate Mab
		
		tpl = lookup(action)	# check if we have a Tilt template
		
		raise "Can't find action #{action}" unless tpl
		
		s = (tpl == true) ? m.capture{m.send(action)} : tpl.render(self, {}) # if so render it
		if /^_/!~action.to_s
			layout_tpl = lookup(:layout)
			if layout_tpl
				b = Proc.new { s }
				s = (layout_tpl == true) ? m.capture{send(:layout){s}} : layout_tpl.render(m, {},&b) # if so render it
			end
		end
		s
	end
  end

  
  module Views
    # Call this inside your Views module to set a default format.
    #
    # For example:
    #
    #   module Foobar::Views
    #     module HTML
    #       # ... etc.
    #     end
    #     default_format :XML
    #   end
    def default_format(m)
      mod = "#{self}::#{m.to_s}".constantize
      mab = self.to_s.gsub('::Views','::Mab').constantize		# @techarch : get the Mab class
      mab.class_eval{include mod}
    end
  end

  module Helpers
    # Overrides Camping's routing helper to make it possible to route RESTful resources.
    #
    # Some usage examples:
    #
    #   R(Kittens)            # /kittens
    #   R(Kittens, 'new')     # /kittens/new
    #   R(Kittens, 1, 'meow') # /kittens/1/meow
    #   R(@kitten)            # /kittens/1
    #   R(@kitten, 'meow')    # /kittens/1/meow
    #   R(Kittens, 'list', :colour => 'black')  # /kittens/list?colour=black
    #
    # The current output format is retained, so if the current <tt>@format</tt> is <tt>:XML</tt>,
    # the URL will be /kittens/1.xml rather than /kittens/1.
    #
    # Note that your controller names might not be loaded if you're calling <tt>R</tt> inside a
    # view module. In that case you should use the fully qualified name (i.e. Myapp::Controllers::Kittens)
    # or include the Controllers module into your view module.
    def R(c, *g)

		cl = c.class.name.split("::").last.pluralize
		app_name = c.class.name.split("::").first
		ctrl_cl = app_name + '::Controllers'				# @techarch : get to the Controllers using the current app
		ctrl = (app_name != 'Class') ? ctrl_cl.constantize : Controllers
		
		if ctrl.constants.include?(cl)				#@techarch updated to use new cl variable
			path = "/#{cl.underscore}/#{c.id}"
			path << ".#{@format.to_s.downcase}" if @format
			path << "/#{g.shift}" unless g.empty?
			self / path
		  elsif c.respond_to?(:restful?) && c.restful?
			base = c.name.split("::").last.underscore
			id_or_action = g.shift  
			if id_or_action.to_s =~ /\d+/			#@techarch needed a to_s after id_or_action to allow pattern matching
			  id = id_or_action
			  action = g.shift
			else
			  action = id_or_action
			end

			path = "/#{base}"
			path << "/#{id}" if id
			path << "/#{action}" if action
			path << ".#{@format.to_s.downcase}" if @format
			
			#@techarch substituted U for u=Rack::Utils
			u=Rack::Utils
			path << "?#{g.collect{|a|a.collect{|k,v| u.escape(k)+"="+u.escape(v)}.join("&")}.join("&")}" unless g.empty? # FIXME: undefined behaviour if there are multiple arguments left
			return path
		else
			_R(c, *g)
		end
	end # def R
  end # module Helpers

  module Controllers
    def self.determine_format(input, env) #:nodoc:
      if input[:format] && !input[:format].empty?
        input[:format].upcase.intern
      elsif env['PATH_INFO'] =~ /\.([a-z]+)$/
        $~[1].upcase.intern
      end
    end

    # Calling <tt>REST "<resource name>"</tt> creates a controller with the
    # appropriate routes and maps your REST methods to standard
    # Camping controller mehods. This is meant to be used in your Controllers
    # module in place of <tt>R <routes></tt>.
    #
    # Your REST class should define the following methods:
    #
    # * create
    # * read(id)
    # * update(id)
    # * destroy(id)
    # * list
    #
    # Routes will be automatically created based on the resource name fed to the
    # REST method. <b>Your class must have the same (but CamelCaps'ed)
    # name as the resource name.</b> So if your resource name is 'kittens',
    # your controller class must be Kittens.
    #
    # For example:
    #
    #   module Foobar::Controllers
    #     class Kittens < REST 'kittens'
    #       # POST /kittens
    #       def create
    #       end
    #
    #       # GET /kittens/(\d+)
    #       def read(id)
    #       end
    #
    #       # PUT /kittens/(\d+)
    #       def update(id)
    #       end
    #
    #       # DELETE /kittens/(\d+)
    #       def destroy(id)
    #       end
    #
    #       # GET /kittens
    #       def list
    #       end
    #     end
    #   end
    #
    # Custom actions are also possible. For example, to implement a 'meow'
    # action simply add a 'meow' method to the above controller:
    #
    #   # POST/GET/PUT/DELETE /kittens/meow
    #   # POST/GET/PUT/DELETE /kittens/(\d+)/meow
    #   def meow(id)
    #   end
    #
    # Note that a custom action will respond to all four HTTP methods
    # (POST/GET/PUT/DELETE).
    #
    # Optionally, you can specify a <tt>:prefix</tt> key that will prepend the
    # given string to the routes. For example, the following will create all
    # of the above routes, prefixed with "/pets"
    # (i.e. <tt>POST '/pets/kittens'</tt>,  <tt>GET '/pets/kittens/(\d+)'</tt>,
    # etc.):
    #
    #   module Foobar::Controllers
    #     class Items < REST 'kittens', :prefix => '/pets'
    #       # ...
    #     end
    #   end
    #
    # Format-based routing similar to that in ActiveResource is also implemented.
    # For example, to get a list of kittens in XML format, place a
    # <tt>GET</tt> call to <tt>/kittens.xml</tt>.
    # See the documentation for the render() method for more info.
    #
    def REST(r, options = {})
      crud = R "#{options[:prefix]}/#{r}/([0-9a-zA-Z]+)/([a-z_]+)(?:\.[a-z]+)?",
        "#{options[:prefix]}/#{r}/([0-9a-zA-Z]+)(?:\.[a-z]+)?",
        "#{options[:prefix]}/#{r}/([a-z_]+)(?:\.[a-z]+)?",
        "#{options[:prefix]}/#{r}(?:\.[a-z]+)?"

      crud.module_eval do
        meta_def(:restful?){true}

        $LOG.debug("Creating RESTful controller for #{r.inspect} using Reststop #{'pull version number here'}") if $LOG

        def get(id_or_custom_action = nil, custom_action =  nil) # :nodoc:
          id = input['id'] if input['id']

          custom_action = input['action'] if input['action']

          if self.methods.include? id_or_custom_action
            custom_action ||= id_or_custom_action
            id ||= nil
          else
            id ||= id_or_custom_action
          end

          id = id.to_i if id && id =~ /^[0-9]+$/

          @format = Reststop::Controllers.determine_format(input, @env)

          begin
            if id.nil? && input['id'].nil?
              custom_action ? send(custom_action) : list
            else
              custom_action ? send(custom_action, id || input['id']) : read(id || input['id'])
            end
          rescue NoMethodError => e
            # FIXME: this is probably not a good way to do this, but we need to somehow differentiate
            #        between 'no such route' vs. other NoMethodErrors
            if e.message =~ /no such method/
              return no_method(e)
            else
              raise e
            end
          rescue ActiveRecord::RecordNotFound => e
            return not_found(e)
          end
        end


        def post(custom_action = nil) # :nodoc:
          @format = Reststop::Controllers.determine_format(input, @env)
          custom_action ? send(custom_action) : create
        end

        def put(id, custom_action = nil) # :nodoc:
          id = id.to_i if id =~ /^[0-9]+$/
          @format = Reststop::Controllers.determine_format(input, @env)
          custom_action ? send(custom_action, id || input['id']) : update(id || input['id'])
        end

        def delete(id, custom_action = nil) # :nodoc:
          id = id.to_i if id =~ /^[0-9]+$/
          @format = Reststop::Controllers.determine_format(input, @env)
          custom_action ? send(custom_action, id || input['id']) : destroy(id || input['id'])
        end

        private
        def _error(message, status_code = 500, e = nil)
          @status = status_code
          @message = message
          begin
            render "error_#{status_code}".intern
          rescue NoMethodError
            if @format.to_s == 'XML'
              "<error code='#{status_code}'>#{@message}</error>"
            else
              out  = "<strong>#{@message}</strong>"
              out += "<pre style='color: #bbb'><strong>#{e.class}: #{e}</strong>\n#{e.backtrace.join("\n")}</pre>" if e
              out
            end
          end
        end

        def no_method(e)
          _error("No controller method responds to this route!", 501, e)
        end

        def not_found(e)
          _error("Record not found!", 404, e)
        end
      end
      crud
    end # def REST
  end # module Controllers
end # module Reststop

module Markaby
  class Builder
    # Modifies Markaby's 'form' generator so that if a 'method' parameter
    # is supplied, a hidden '_method' input is automatically added. 
    def form(*args, &block)
      options = args[0] if args && args[0] && args[0].kind_of?(Hash)
      inside = capture &block
      
      if options && options.has_key?(:method)
        inside = input(:type => 'hidden', :name => '_method', :value => options[:method]) +
          inside
        if options[:method].to_s === 'put' || options[:method].to_s == 'delete'
          options[:method] = 'post'
        end
      end
      
      tag!(:form, options || args[0]) {inside}
    end
  end
end
